### PR TITLE
TYP: annotations in core.groupby

### PR DESCRIPTION
**Pull request by jbrockmendel**
I'm still seeing a couple of mypy complaints, suggestions @simonjayhawkins ?

### CODE REVIEW

File: pandas/core/groupby/generic.py [hunk 1]
```diff
@@ -1638,7 +1638,7 @@ def _get_data_to_aggregate(self) -> BlockManager:
         else:
             return obj._mgr
 
-    def _insert_inaxis_grouper_inplace(self, result):
+    def _insert_inaxis_grouper_inplace(self, result: DataFrame):
```
Comment (simonjayhawkins) [inline pandas/core/groupby/generic.py:1641]: for functions other than \_\_init__ you can add None as the return type.

File: pandas/core/groupby/groupby.py [hunk 2]
```diff
@@ -734,7 +734,7 @@ def pipe(self, func, *args, **kwargs):
 
     plot = property(GroupByPlot)
 
-    def _make_wrapper(self, name):
+    def _make_wrapper(self, name: str) -> Callable:
```
Comment (simonjayhawkins) [inline pandas/core/groupby/groupby.py:737]: add type parameters for Callable (if you can)

**Comment by simonjayhawkins** [comment]
> I'm still seeing a couple of mypy complaints, suggestions @simonjayhawkins ?

pandas\core\groupby\categorical.py:101: error: "CategoricalIndex" has no attribute "set_categories"; maybe "categories"?  [attr-defined]
pandas\core\groupby\categorical.py:104: error: "CategoricalIndex" has no attribute "add_categories"; maybe "categories"?  [attr-defined]

These are because mypy is a static type checker and these methods are dynamically added to CategoricalIndex

in `DatetimeIndex`, tz_localize and to_period were added statically https://github.com/pandas-dev/pandas/pull/34630#discussion_r436402735

**Comment by simonjayhawkins** [comment]
error: Argument 1 to "map" has incompatible type overloaded function; expected "Callable[[object], Iterator[_T]]"  [arg-type]

not sure on this one.

mypy is happy with
```
        for name, lev, in_axis in zip(
            reversed(self.grouper.names),
            reversed(self.grouper.get_group_levels()),
            reversed([grp.in_axis for grp in self.grouper.groupings]),
        ):
```
which I think is easier to grok anyhow.

**Comment by simonjayhawkins** [comment]
see https://github.com/python/mypy/issues/6811#issuecomment-491332516

>As for the originally reported bug, it also repros with this simpler code:

>```
>def get_common_index(*names: str) -> int:
>    x = map(reversed, names)
>    return 0
>```

File: pandas/core/groupby/generic.py [hunk 3]
```diff
@@ -1848,6 +1848,7 @@ def nunique(self, dropna: bool = True):
             ],
             axis=1,
         )
+        assert isinstance(results, DataFrame)  # for mypy
```
Comment (simonjayhawkins) [inline pandas/core/groupby/generic.py:1851]: if you use cast here, then once we adopt Literal and concat can be overloaded on axis value, then with `warn_redundant_casts` we can remove

File: pandas/core/groupby/grouper.py [hunk 4]
```diff
@@ -568,7 +568,9 @@ def codes(self) -> np.ndarray:
     @cache_readonly
     def result_index(self) -> Index:
         if self.all_grouper is not None:
-            return recode_from_groupby(self.all_grouper, self.sort, self.group_index)
+            group_idx = self.group_index
+            assert isinstance(group_idx, CategoricalIndex)  # set in __init__
+            return recode_from_groupby(self.all_grouper, self.sort, group_idx)
```
Comment (simonjayhawkins) [inline pandas/core/groupby/grouper.py:573]: is this needed.

are the type annotations for `def group_index(self) -> Index` and `_group_index: Optional[Index] = None` correct.

in \_\_init__

```
self._group_index = CategoricalIndex(..
```
or 
```
(
                self.grouper,
                self._codes,
                self._group_index,
            ) = index._get_grouper_for_level(self.grouper, level)

```
so self._group_index  can only be CategoricalIndex or None and group_index can only be CategoricalIndex ?

File: pandas/core/groupby/generic.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/groupby/generic.py:1851]: neat, will do

File: pandas/core/groupby/grouper.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/core/groupby/grouper.py:573]: IIRC its the self.all_grouper check a few lines up that ensures we have a CategoricalIndex here

File: pandas/core/groupby/groupby.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/groupby/groupby.py:737]: its going to end up being `wrapper(self, *args, **kwargs) -> FrameOrSeriesUnion`.  Is that `Callable[..., FrameOrSeriesUnion]`?

File: pandas/core/groupby/grouper.py (hunk 4 above)
Comment (simonjayhawkins) [reply pandas/core/groupby/grouper.py:573]: self.all_grouper is not None when is_categorical_dtype(self.grouper) so I don't think that narrows it. (but I may need to trace through further)

File: pandas/core/groupby/groupby.py (hunk 2 above)
Comment (simonjayhawkins) [reply pandas/core/groupby/groupby.py:737]: sure

**Review by simonjayhawkins — APPROVED** [review]
Thanks @jbrockmendel generally lgtm.

**Comment by simonjayhawkins** [comment]
Thanks @jbrockmendel

**Pull request merged**